text/0481-component-templates-co-location.md

```diff
@@ -0,0 +1,599 @@
+like in your app, the following commands give a close approximation:
+
+```bash
+$ rsync --archive --remove-source-files app/templates/components/ app/components/
+$ rm -rf app/templates/components/
+```
+
+Of course, the resulting output won't "work", but it can be useful for getting
+a sense of what it's like to work with the new layout on editors, Github, etc.
+
+### Generator
+
+We propose to make some updates to the components generator to accompany this
+change.
+
+1. It should accept a `--component-class` option. This controls which base
+   class is used for that component and whether native classes are used. The
+   legal values for this option are `@glimmer/component` (aliased as `-gc`),
+   `@ember/component` (aliased as `-cc`), `@ember/component/template-only`
+   (aliased as `-tc`) or an empty string (aliased as `--no-component-class` and
+   `-nc`).
+
+   The latter two differ in that `@ember/component/template-only` would
+   generate an explicit JavaScript file with `templateOnlyComponent()` as the
+   default export, which is useful for addons, whereas `--no-component-class`
+   would skip generating a JavaScript file altogether.
+
+   This option may be extended in the future to allow other custom components
+   to provide their own blueprints, but for now, passing anything other than
+   the allowed values will result in an error.
+
+2. It should accept a `--component-structure` option.
+
+   When this option is set to `flat` (aliased as `-fs`), the component's JavaScript and template files will both be generated at the root of
+   `{app,addon}/components`.
+
+   When this option is set to `nested` (aliased as `-ns`), a folder will be
+   generated for the component in `{app,addon}/components`, and the component's
+   JavaScript and template files will be generated as `index.{js,hbs}` inside
+   the folder.
+
+   When this option is set to `classic` (aliased as `-cs`), the component's
+   template file will be generated in `{app,addon}/templates/components`. For
+   addons, when used with `--component-class=@ember/component`, this will also
+   emit the `layout` property workaround.
+
+3. Thses options will default to `--component-class=@ember/component` and
+   `--component-structure=classic` for backwards compatibility.
+
+   However, the default values can be overridden in `.ember-cli` as usual, and
+   teams are encouraged to do so as they see fit. Due to a limitation in how
+   the system works, the names for these options are chosen such that they are
+   unlikely to conflict with options on other `ember` commands, which is why
+   they are a bit verbose.
+
+   For Octane apps, the default app blueprint will include a `.ember-cli` file
+   that defaults to `--no-component-class` and `--component-structure=flat`.
+   The guides and documentation will assume these settings going forward.
+
+## How we teach this
+
+As mentioned above, the learning team will update the guides and API docs to
```

Comment:
Suggested change to this first paragraph:

> As mentioned above, we will update the learning resources to assume the "flat" co-located layout. Throughout the Ember Guides, the Tutorials, and the CLI Addon Tutorial, we would update all Component Handlebars code sample filenames. The prose describing the location of files would also need to change. The API documentation will describe the full set of file layout options supported in the major version of Ember. A section should also be added to the CLI guides, which describes how to properly import components from addons that have a mix of layout types. We will not cover blending file layouts in the Ember Guides, since they represent the happy path for an app, but we could link to the CLI Guides explanation.

Reply:
👍 I applied the suggestions manually.

I continue to have doubts that it would be appropriate to cover these kind of things in the API docs, although I understand the motivation. Specifically, there is no natural place to cover abstract things like these in API docs – which must be attached to a concrete item, like a class or a function. Things like this (and angle/curly invocation styles, which was the last time it came up) are not a good fit for that because they apply to the abstract concept of components in general, not a specific class like `Ember.Component` (which is where they currently live as a compromise). I believe @wycats shares my concerns too.

However, we can stick with the current policy for the purpose of this RFC and discuss options outside of this thread.